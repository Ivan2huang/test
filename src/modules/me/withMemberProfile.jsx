/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import axios from 'axios';
import {
  ACCESS_TOKEN,
  USER_ID,
  LOGGED_IN,
  CLIENT_ID,
  routesWithoutFetchData,
  CUSTOMTOKEN_HEADER,
} from '../../constants/auth';
import { SERVICE_IDENTIFIERS } from '../../constants/api';

import { updateMemberProfile } from './action';
import { transformMemberDetails } from './api';
import logger from '../../helpers/logger';

const dev = process.env.NODE_ENV === 'development';

const getMemberProfile = (accessToken, memberId, clientId, ctx) => {
  logger.log('---getMemberProfile -> START---');

  const url = `${
    process.env.MEMBER_SERVICE
  }${SERVICE_IDENTIFIERS.memberService.replace(
    '/member-service/',
    '',
  )}/${clientId}/users/${memberId}/profile`;
  const authorizationHeader = {
    Authorization: `Bearer ${accessToken}`,
  };
  if (!dev) {
    authorizationHeader[CUSTOMTOKEN_HEADER] =
      ctx.req.headers[CUSTOMTOKEN_HEADER] || '';
  }
  const options = {
    method: 'GET',
    headers: authorizationHeader,
  };

  console.log('withMemberProfile.jsx getMemberProfile url=', url);
  console.log('withMemberProfile.jsx getMemberProfile options=', options);

  return axios(url, options)
    .then(response => {
      const { data } = response;
      logger.log('---getMemberProfile -> END---');
      /**
       * 用 huyduong@mailinator.com  和  P@ssw0rd  登录后可拿到 的 data :
       * {
            clientId: 'cxaref-dev',
            memberId: '147',
            fullName: 'Huy Duong',
            hasLoggedIn: true,
            firstName: 'Huy',
            lastName: 'Duong',
            workEmail: 'huyduong@mailinator.com',
            externalId: 'huyduong',
            gender: 'Male',
            dateOfBirth: '1990-09-10T00:00:00',
            lastLogin: '2021-03-19T10:59:14.813293',
            role: 'Employee',
            category: 'Tier 1A',
            relationshipToEmployee: 'Self',
            relationshipCategory: 'Self',
            preferredLocale: 'en-UK',
            currency: 'HKD',
            certificateNumber: 'huyduong',
            dependentNumber: '3',
            relationships: [],
            isEmailVerified: true,
            isTermsAccepted: true,
            isEdmOptedOut: true,
            isSelfRegisteredUser: false,
            status: 'None',
            accountStatus: 'Active',
            policyNumber: 'huyduong',
            preferedEmail: 'huyduong@mailinator.com',
            nationality: 'Hong Kong',
            identificationNumber: 'huyduong',
            identificationType: 'nana',
            membershipNumber: 'huyduong'
          }
       */
      console.log('withMemberProfile.jsx getMemberProfile data=', data);
      return data;
    })
    .catch(error => {
      logger.error('---getMemberProfile -> ERROR---', error);
    });
};

const getFromCookieOrLocals = ctx => {
  if (ctx && ctx.req) {
    const accessToken =
      ctx.req.cookies[ACCESS_TOKEN] || ctx.res.locals.accessToken;
    const userId = ctx.req.cookies[USER_ID] || ctx.res.locals.userId;
    const isLoggedIn = ctx.req.cookies[LOGGED_IN] || ctx.res.locals.isLoggedIn;
    const preferredLocale = ctx.req.cookies.lang || ctx.res.locals.lang;
    const clientId =
      ctx.req.cookies[CLIENT_ID] ||
      ctx.res.locals.clientId;
    return { isLoggedIn, userId, accessToken, preferredLocale, clientId };
  }
  return {};
};

const withMemberProfile = WrappedComponent =>
  class extends Component {
    static async getInitialProps({ ctx, ...rest }) {
      console.log('withMemberProfile.jsx getInitialProps');
      let memberProfile = {};
      const {
        isLoggedIn,
        userId,
        accessToken,
        preferredLocale,
        clientId,
      } = getFromCookieOrLocals(ctx);

      const shouldCallMemberProfile =
        isLoggedIn &&
        userId &&
        accessToken &&
        routesWithoutFetchData.indexOf(ctx.req.path) === -1;

      if (shouldCallMemberProfile) {
        memberProfile =
          (await getMemberProfile(accessToken, userId, clientId, ctx)) || {};
        ctx.store.dispatch(
          updateMemberProfile(transformMemberDetails(memberProfile)),
        );
      }

      const componentProps =
        WrappedComponent.getInitialProps &&
        (await WrappedComponent.getInitialProps({ ...ctx, ...rest }));

      return {
        memberProfile,
        ...componentProps,
        preferredLocale,
      };
    }

    render() {
      console.log('withMemberProfile.jsx render');

      return <WrappedComponent {...this.props} />;
    }
  };

export default withMemberProfile;
