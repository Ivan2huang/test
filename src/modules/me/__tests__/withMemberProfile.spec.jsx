import React from 'react';
import { render } from '@testing-library/react';

import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import withMemberProfile from '../withMemberProfile';

jest.mock('next/router');

const mockedDispatch = jest.fn();
jest.mock('../../../redux/store', () => ({
  getStore: () => ({
    dispatch: mockedDispatch,
  }),
}));

jest.mock('../../../constants/auth', () => ({
  ACCESS_TOKEN: 'access_token',
  LOGGED_IN: 'logged_in',
  USER_ID: 'user_id',
  routesWithoutFetchData: ['/error'],
}));

jest.mock('../../../helpers/paths', () => ({
  common: {
    error: '/error',
  },
}));

jest.mock('../../../helpers/logger');

const request = new MockAdapter(axios);

const ctxWithoutToken = {
  req: {
    cookies: { id_token: null },
    headers: {},
  },
  res: {
    redirect: jest.fn(),
    locals: {},
  },
  store: {
    dispatch: mockedDispatch,
  },
};
const ctxWithToken = {
  req: {
    cookies: {
      id_token: 'test_token',
      access_token: 'test_access_token',
      user_id: 'user_id_1',
      logged_in: true,
      lang: 'en-HK',
    },
    headers: {},
  },
  res: {
    redirect: jest.fn(),
    locals: {
      access_token: 'test_access_token',
    },
  },
  store: {
    dispatch: mockedDispatch,
  },
};

afterEach(() => {
  jest.clearAllMocks();
});

const TestComponent = () => <div>Test Component </div>;
TestComponent.getInitialProps = () => Promise.resolve({});

describe('with MemberProfile Provider', () => {
  it('should match snapshot', async () => {
    const Component = withMemberProfile(TestComponent);

    const { container } = render(
      <Component memberProfile={{ relationships: [] }} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should return empty memberProfile when token is missing from cookie', async () => {
    const Component = withMemberProfile(TestComponent);

    const memberProfile = await Component.getInitialProps({
      ctx: ctxWithoutToken,
    });

    expect(memberProfile).toEqual({ memberProfile: {} });
  });

  it('should return empty memberProfile when on client side', async () => {
    const Component = withMemberProfile(TestComponent);

    const memberProfile = await Component.getInitialProps({
      ctx: {},
    });

    expect(memberProfile).toEqual({ memberProfile: {} });
  });

  it('should return member profile when valid token is passed and api return success', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(200, {});

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should call disabled wallet with valid balance', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(200, { member: { balance: 0 } });

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should return member profile when valid token is passed and api return success even wallet error', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(404, {
        errors: [{ messageKey: '' }],
      });

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should return member profile when valid token is passed and api return success even invalid wallet error', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(404, null);

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should return member profile when valid token is passed and api return success even single wallet error', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(404, { code: 'test' });

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should return member profile and disable wallet when valid token is passed and api return success', async () => {
    const Component = withMemberProfile(TestComponent);
    request
      .onGet(new RegExp('profile$'))
      .reply(200, { preferredLocale: 'en-HK' })
      .onGet(new RegExp('/balance\\?includeDependents'))
      .reply(402, {
        errors: [{ messageKey: 'INTERNAL_SERVER_ERROR' }],
      });

    const actual = await Component.getInitialProps({ ctx: ctxWithToken });
    const expected = {
      memberProfile: { preferredLocale: 'en-HK' },
      preferredLocale: 'en-HK',
    };

    expect(actual).toEqual(expected);
  });

  it('should return member profile when valid token is passed and api returns Unauthorized', async () => {
    const Component = withMemberProfile(TestComponent);
    request.onGet(new RegExp('profile$')).reply(401, {});

    const memberProfile = await Component.getInitialProps({
      ctx: ctxWithToken,
    });

    expect(memberProfile).toEqual({
      memberProfile: {},
      preferredLocale: 'en-HK',
    });
  });

  it('should not redirect to error page when current request is from error page', async () => {
    const Component = withMemberProfile(TestComponent);
    request.onGet(new RegExp('profile$')).reply(401, {});

    ctxWithToken.req.path = '/error';
    const memberProfile = await Component.getInitialProps({
      ctx: ctxWithToken,
    });

    expect(ctxWithToken.res.redirect).not.toHaveBeenCalled();
    expect(memberProfile).toEqual({
      memberProfile: {},
      preferredLocale: 'en-HK',
    });
  });
});
