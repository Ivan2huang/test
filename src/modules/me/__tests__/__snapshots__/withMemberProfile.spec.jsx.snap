// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with MemberProfile Provider should match snapshot 1`] = `
<div>
  <div>
    Test Component 
  </div>
</div>
`;
